pkg/typesystem/typesystem_test.go

```diff
@@ -12,6 +12,423 @@ import (
 	"github.com/openfga/openfga/pkg/testutils"
 )
 
+type RelationDetails struct {
+	hasEntrypoints bool
+	hasCycle       bool
+}
+
+func TestHasEntrypoints(t *testing.T) {
+	tests := map[string]struct {
+		model         string
+		inputType     string
+		inputRelation string
+		expectError   string
+		expectDetails *RelationDetails
+	}{
+		`undefined_input_relation`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [folder]`,
+			inputType:     "document",
+			inputRelation: "unknown",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`this_has_entrypoints_through_user`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_user_wildcard`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document:*]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_has_entrypoints_through_userset`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type org
+				relations
+					define member: [user]
+			type folder
+				relations
+					define parent: [org#member]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_first`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define parent: [user, folder#parent]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		`this_with_two_assignable_types_has_entrypoints_through_second`: {
+			model: `
+			model
+				schema 1.1
+			type user
+			type folder
+				relations
+					define editor: [user]
+					define parent: [folder#parent, folder#editor]`,
+			inputType:     "folder",
+			inputRelation: "parent",
+			expectDetails: &RelationDetails{true, false},
+		},
+		// TODO fix
+		//`this_has_no_entrypoints_because_type_unknown_is_not_defined`: {
```

Comment:
If I may suggestion a different approach, what do you think about actually uncommenting all of this and instead in the t.Run check for the name of the tests and skip if it's this one? Same applies to the other commented out tests. Our goal is to skip them but we know that the setup is good, so let's not let that good work just be commented out. 

```
	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			switch name {
			case "this_has_no_entrypoints_because_type_unknown_is_not_defined",
				"issue_1260_parallel_edges_mean_entrypoints":
				t.Skip("not yet fixed")
			}

			model := testutils.MustTransformDSLToProtoWithID(test.model)
			ts := New(model)
			inputRelation, _ := ts.GetRelation(test.inputType, test.inputRelation)

			rewrite := inputRelation.GetRewrite()
			hasEntrypoints, hasCycle, err := hasEntrypoints(ts.GetAllRelations(), test.inputType, test.inputRelation, rewrite, map[string]map[string]struct{}{})

			if test.expectError != "" {
				require.ErrorContains(t, err, test.expectError)
				return
			} 
			require.NoError(t, err)
			require.Equal(t, test.expectDetails.hasEntrypoints, hasEntrypoints, "unexpected value for hasEntrypoints")
			require.Equal(t, test.expectDetails.hasCycle, hasCycle, "unexpected value for hasCycle")
			
		})
	}
```

Reply:
That feels like a lot of setup for something very short-lived :) i already uncommented these tests in https://github.com/openfga/openfga/pull/1405